### PR TITLE
Remove `_debugWillReattachChildren` assertions from `_TableElement`

### LINKED ISSUE #31473: Table widget doesn't work with Hero animations.
## Steps to reproduce

Run the following app, and tap on the text.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: Home()));

class Home extends StatelessWidget {
  const Home({
    Key key,
  }) : super(key: key);

  void startHeroTransition(BuildContext context) {
    Navigator.of(context).push(MaterialPageRoute(
      builder: (BuildContext context) => Center(child: HeroText()),
    ));
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: GestureDetector(
        onTap: () => startHeroTransition(context),
        child: Table(
          children: [
            TableRow(
              children: [
                HeroText(),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class HeroText extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Hero(
      tag: 'TAG',
      child: Material(child: Text('Tap here to reproduce the bug')),
    );
  }
}
```

## Expected behavior

The text should animate via a hero transition to the center of the screen, with no exceptions in the logs.

## Actual behavior

```
I/flutter (23398): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter (23398): The following assertion was thrown building Hero(tag: TAG, state: _HeroState#f6027):
I/flutter (23398): 'package:flutter/src/widgets/table.dart': Failed assertion: line 286 pos 12: '() {
I/flutter (23398):       if (_debugWillReattachChildren)
I/flutter (23398):         return true;
I/flutter (23398):       for (Element forgottenChild in _forgottenChildren) {
I/flutter (23398):         if (forgottenChild.renderObject == child)
I/flutter (23398):           return true;
I/flutter (23398):       }
I/flutter (23398):       return false;
I/flutter (23398):     }()': is not true.
I/flutter (23398): 
I/flutter (23398): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter (23398): more information in this error message to help you determine and fix the underlying cause.
I/flutter (23398): In either case, please report this assertion by filing a bug on GitHub:
I/flutter (23398):   https://github.com/flutter/flutter/issues/new?template=BUG.md
I/flutter (23398): 
I/flutter (23398): When the exception was thrown, this was the stack:
I/flutter (23398): #2      _TableElement.removeChildRenderObject (package:flutter/src/widgets/table.dart:286:12)
I/flutter (23398): #3      RenderObjectElement.detachRenderObject (package:flutter/src/widgets/framework.dart:4744:36)
I/flutter (23398): #4      Element.detachRenderObject.<anonymous closure> (package:flutter/src/widgets/framework.dart:2867:13)
I/flutter (23398): #5      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3770:14)
I/flutter (23398): #6      Element.detachRenderObject (package:flutter/src/widgets/framework.dart:2866:5)
I/flutter (23398): #7      Element.detachRenderObject.<anonymous closure> (package:flutter/src/widgets/framework.dart:2867:13)
I/flutter (23398): #8      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3770:14)
I/flutter (23398): #9      Element.detachRenderObject (package:flutter/src/widgets/framework.dart:2866:5)
I/flutter (23398): #10     Element.detachRenderObject.<anonymous closure> (package:flutter/src/widgets/framework.dart:2867:13)
I/flutter (23398): #11     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:3770:14)
I/flutter (23398): #12     Element.detachRenderObject (package:flutter/src/widgets/framework.dart:2866:5)
I/flutter (23398): #13     Element.deactivateChild (package:flutter/src/widgets/framework.dart:2996:11)
I/flutter (23398): #14     Element.updateChild (package:flutter/src/widgets/framework.dart:2761:7)
I/flutter (23398): #15     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3750:16)
I/flutter (23398): #16     Element.rebuild (package:flutter/src/widgets/framework.dart:3565:5)
I/flutter (23398): #17     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2278:33)
I/flutter (23398): #18     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:700:20)
I/flutter (23398): #19     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:286:5)
I/flutter (23398): #20     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1012:15)
I/flutter (23398): #21     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:952:9)
I/flutter (23398): #22     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:864:5)
I/flutter (23398): #26     _invoke (dart:ui/hooks.dart:219:10)
I/flutter (23398): #27     _drawFrame (dart:ui/hooks.dart:178:3)
I/flutter (23398): (elided 5 frames from class _AssertionError and package dart:async)
I/flutter (23398): ════════════════════════════════════════════════════════════════════════════════════════════════════
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

Calling `insertChildRenderObject` and `removeChildRenderObject` outside of `update` or `mount` in `_TableElement` should be allowed. 

Consider the following minimal reproducible example:
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      home: Center(
        child: SizedBox(
          width: 100,
          height: 100,
          child: Table(children: [TableRow(children: [W()])]),
        ),
      ),
    ),
  );
}

class W extends StatefulWidget {
  const W({ Key key }) : super(key: key);

  @override
  _State createState() => _State();
}

class _State extends State<W> {
  bool _toggle = true;
  @override
  Widget build(BuildContext context) {
    return _toggle
        ? GestureDetector(onTap: () => setState(() => _toggle = !_toggle), child: Text('Tap me'))
        : SizedBox();
  }
}
```

### Sequence of events after tapping 'Tap me'
1. The `onTap` callback is called, `_State.context` is  marked dirty. 
2. In the next build pass, because `GestureDetector` was replaced by a widget of a different type,  `_State.context.updateChild` method calls `_State.context.deactivateChild`.
3. The child's renderObject asks `_TableElement` to detach it from `_TableElement.renderObject`, calling `_TableElement.removeChildRenderObject`.
4. The assertion in `_TableElement.removeChildRenderObject` fails.

In the above call path only `removeChildRenderObject` was called on a `_TableElement` instance, pretty much everything else is happening in the descendant elements which we have no knowledge of in the implementation of `Table`. So it's probably easier to just remove the asserts than figuring out a way to keep them.

## Related PR

`_debugWillReattachChildren` wasn't brought up in #2961's discussion.

## Related Issues

Fixes #31473

## Tests

I added the following tests:

-   'Does not crash if a child RenderObject is replaced by another RenderObject of a different type',

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/table_test.dart [hunk 1]
```diff
@@ -18,6 +18,28 @@ class TestStatefulWidgetState extends State<TestStatefulWidget> {
...
+  bool toggle = true;
+
+  void toggleMe() {
+    setState(() { toggle = !toggle; });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return toggle
+    ? const SizedBox()
```
Comment (goderbauer) [inline packages/flutter/test/widgets/table_test.dart]: nit: indentation is off here (this could probably just all be in one line)

**Review by goderbauer — COMMENTED** [review]
Did you checkin with @Hixie, who wrote the original code, to figure out if he remembers why the assert was added in the first place? There may have been a good reason for them to be there in the first place.

File: packages/flutter/test/widgets/table_test.dart [hunk 2]
```diff
@@ -819,5 +837,32 @@ void main() {
...
   });
 
+  // Regression test for https://github.com/flutter/flutter/issues/31473.
+  testWidgets(
+    'Does not crash if a child RenderObject is replaced by another RenderObject of a different type',
+    (WidgetTester tester) async {
+      await tester.pumpWidget(
+        Directionality(
+          textDirection: TextDirection.ltr,
+          child: Table(children: const <TableRow> [TableRow(children: <Widget> [TestChildWidget()])]),
```
Comment (Hixie) [inline packages/flutter/test/widgets/table_test.dart]: trivial nit: no space between `>` and `[`

**Comment by Hixie** [comment]
LGTM

yeah i think this assert is just bogus because the table doesn't necessarily know if some widget below it gets removed and readded with another type.

**Pull request merged**